src/Database/Beam/AutoMigrate/Types.hs

```diff
@@ -172,32 +199,75 @@ newtype DbEnum a
+-- of the table where the foreign key is actually defined (in terms of references).
+-- The set stores a (fk_column, pk_column) correspondence.
+data ForeignKey = ForeignKey
+  TableName
+  (Set (ColumnName, ColumnName))
+  deriving (Eq, Ord, Show, Generic)
+
+instance NFData ForeignKey
 
-instance Monoid Table where
-  mempty = Table mempty mempty
+data ForeignKeyConstraintOptions = ForeignKeyConstraintOptions
+  { foreignKeyConstraintName :: Maybe ConstraintName -- ^ The Maybe indicates not that the constraint might not have a name, but that we aren't obligated it to name it immediately.
+  , onDelete :: ReferenceAction {- onDelete -}
+  , onUpdate :: ReferenceAction {- onUpdate -}
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+instance Default ForeignKeyConstraintOptions where
+  def = ForeignKeyConstraintOptions Nothing NoAction NoAction
 
-type ConstraintName = Text
+instance Semigroup ForeignKeyConstraintOptions where
+  ForeignKeyConstraintOptions n1 d1 u1 <> ForeignKeyConstraintOptions n2 d2 u2 = ForeignKeyConstraintOptions
+    (n1 <|> n2)
+    (const d1 d2)
+    (const u1 u2)
 
-data TableConstraint
-  = -- | This set of 'Column's identifies the Table's 'PrimaryKey'.
-    PrimaryKey ConstraintName (Set ColumnName)
-  | -- | This set of 'Column's identifies a Table's 'ForeignKey'. This is usually found in the 'tableConstraints'
-    -- of the table where the foreign key is actually defined (in terms of 'REFERENCES').
-    -- The set stores a (fk_column, pk_column) correspondence.
-    ForeignKey ConstraintName TableName (Set (ColumnName, ColumnName)) ReferenceAction {- onDelete -} ReferenceAction {- onUpdate -}
-  | Unique ConstraintName (Set ColumnName)
+instance NFData ForeignKeyConstraintOptions
+
+data Unique = Unique (Set ColumnName)
   deriving (Show, Eq, Ord, Generic)
 
-instance NFData TableConstraint
+instance NFData Unique
+
+data UniqueConstraintOptions = UniqueConstraintOptions
+  { uniqueConstraintName :: Maybe ConstraintName -- ^ The Maybe indicates not that the constraint might not have a name, but that we aren't obligated it to name it immediately.
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+-- | suitable default options which allows postgres to choose a constraint name
+instance Default UniqueConstraintOptions where
+  def = UniqueConstraintOptions Nothing
+
+instance Semigroup UniqueConstraintOptions where
+  UniqueConstraintOptions n1 <> UniqueConstraintOptions n2 = UniqueConstraintOptions
+    (n1 <|> n2)
+
+instance NFData UniqueConstraintOptions
 
-data ColumnConstraint
-  = NotNull
-  | Default Text {- the actual default -}
+-- | we will treat default constraint expressions specially in the case that a column is associated with a sequence.
+--
+-- We only ever represent two kinds of sequence, either the sequence is owned by a column which has a default of nextval(sequence); or else the sequence is not owned, and must have a name.
+data DefaultConstraint
+  = DefaultExpr Text
+  | Autoincrement (Maybe SequenceName)
```

Comment:
Why does it make sense to represent Autoincrement as different from Default "nextval(whatever)"?